src/deployer/Deployer.ts

```diff
@@ -47,10 +48,91 @@ export class Deployer {
     return adapter.toInstance(contract, contractAddress, parameters);
   }
 
-  public async deployed<T, A = T, I = any>(
-    contract: Instance<A, I> | (T extends Truffle.Contract<I> ? T : never),
-    contractIdentifier?: string,
-  ): Promise<I> {
+  public async deployERC1967Proxy<A, I = any>(
+    implementationFactory: Instance<A, I>,
+    argsOrParameters: OverridesAndLibs | TypedArgs<A> = [] as TypedArgs<A>,
+    parameters: OverridesAndLibs = {},
+  ) {
+    return this.deployProxy(
+      implementationFactory,
+      "ERC1967Proxy",
+      (implementationAddress) => [implementationAddress, "0x"],
+      argsOrParameters,
+      parameters,
+    );
+  }
+
+  public async deployTransparentUpgradeableProxy<A, I = any>(
+    implementationFactory: Instance<A, I>,
+    proxyAdmin: string,
+    argsOrParameters: OverridesAndLibs | TypedArgs<A> = [] as TypedArgs<A>,
+    parameters: OverridesAndLibs = {},
+  ) {
+    if (proxyAdmin === ZeroAddress) {
+      throw new MigrateError("Proxy admin cannot be the zero address");
+    }
+
+    return this.deployProxy(
+      implementationFactory,
+      "TransparentUpgradeableProxy",
+      (implementationAddress) => [implementationAddress, proxyAdmin, "0x"],
+      argsOrParameters,
+      parameters,
+    );
+  }
+
+  public async deployProxy<A, I = any>(
+    implementationFactory: Instance<A, I>,
+    proxyFactoryName: string,
+    proxyArgs: (implementationAddress: string) => any[],
+    argsOrParameters: OverridesAndLibs | TypedArgs<A> = [] as TypedArgs<A>,
+    parameters: OverridesAndLibs = {},
+  ) {
+    const adapter = Deployer.resolveAdapter(this._hre, implementationFactory);
+
+    let implementationArgs: TypedArgs<A> = [] as any;
+    let implementationParameters: OverridesAndLibs = {};
+
+    if (argsOrParameters && Array.isArray(argsOrParameters)) {
+      implementationArgs = argsOrParameters;
+    } else if (argsOrParameters && typeof argsOrParameters === "object") {
+      implementationParameters = argsOrParameters;
+    } else {
+      implementationParameters = parameters;
+    }
+
+    let instanceName = adapter.getContractName(implementationFactory, parameters);
+
+    const implementation = (await this.deploy(implementationFactory, implementationArgs, {
+      ...implementationParameters,
+      name: `${instanceName} implementation`,
+    })) as any;
+    VerificationProcessor.saveVerificationFunction({
+      contractAddress: await implementation.getAddress(),
+      contractName: instanceName,
+      constructorArguments: implementationArgs,
+      chainId: Number(await getChainId()),
+    });
+
+    const { ethers, artifacts } = await import("hardhat");
+    const proxyFactory = await ethers.getContractFactory(proxyFactoryName);
```

Comment:
What if there is no such proxy in the project?

Reply:
It will throw an error